pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+``Callable`` type because the vast majority of callbacks can be correctly
+described by the existing ``typing.Callable`` semantics:
+
+- Positional parameters: By far the most important case to handle well
+  is simple callable types with positional parameters, such as
+  ``(int, str) -> bool``
+- ParamSpec and Concatenate: The next most important feature is good
+  support for PEP 612 ``ParamSpec`` and ``Concatenate`` types like
+  ``(**P) -> bool`` and ``(int, **P) -> bool``. These are common
+  primarily because of the heavy use of decorator patterns in python
+  code.
+- TypeVarTuples: The next most important feature, assuming PEP 646 is
+  accepted, is for unpacked types which are common because of cases
+  where a wrapper passes along ``*args`` to some other function.
+
+Features that other, more complicated proposals would support account
+for fewer than 2% of the use cases we found. These are already
+expressible using Callback Protocols, and since they are uncommon we
+decided that it made more sense to move forward with a simpler syntax.
+
+Extended Syntax Supporting Named and Optional Arguments
+-------------------------------------------------------
+
+Another alternative was for a compatible but more complex syntax that
+could express everything in this PEP but also named, optional, and
+variadic arguments. In this “extended” syntax proposal the following
+types would have been equivalent::
+
+    class Function(typing.Protocol):
+        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
+            ...
+
+    Function = (int, y: float, *, z: bool = ..., **kwargs: str) -> bool
+
+Advantages of this syntax include: - Most of the advantages of the
+proposal in this PEP (conciseness, PEP 612 support, etc) -
+Furthermore, the ability to handle named, optional, and variadic
+arguments
+
+We decided against proposing it for the following reasons:
+
+- The implementation would have been more difficult, and usage stats
+  demonstrate that fewer than 3% of use cases would benefit from any
+  of the added features.
+- The group that debated these proposals was split down the middle
+  about whether these changes are even desirable:
+
+  - On the one hand, they make callable types more expressive. On the
+    other hand, they could easily confuse users who have not read the
+    full specification of callable type syntax.
+  - We believe the simpler syntax proposed in this PEP, which
+    introduces no new semantics and closely mimics syntax in other
+    popular languages like Kotlin, Scala, and TypesScript, is much
+    less likely to confuse users.
+
+- We intend to implement the current proposal in a way that is
+  forward-compatible with the more complicated extended syntax. If the
+  community decides after more experience and discussion that we want
+  the additional features, they should be straightforward to propose
+  in the future.
+- We realized that because of overloads, it is not possible to replace
```

Comment:
I wonder if "because of overloads" requires a complete example (e.g. an overloaded function, the observation that its type can't be expressed using current function types, and the conclusion that it must be done using a Callback Protocol). It might also be useful to explain why you can't just replace the overload with a union of function types. (Hm. Why couldn't you?)

Reply:
+1 on a concrete example.

> It might also be useful to explain why you can't just replace the overload with a union of function types. (Hm. Why couldn't you?)

Overloads correspond more to an intersection of callable types. If we had a union `f: (int) -> str | ((list[int]) -> None)`, then we wouldn't be able to pass an argument to just one of the union items: `f(42)` would error because it's not a `list[int]`; `f([1, 2])` would error because it's not an `int`.

Reply:
```
- No replacement for callback protocol:
```